File: app/js/engine/building-type.ts
/// <reference path="terrain-type.ts" />

module Miner {
  export interface BuildingType {
    id: number;
    name: string;
    cost: number;
    validTerrain: TerrainType;
    constructionDays: number;
    oreProduction: number;
    miningCapacity: number;
    techCapacity: number;
    opsCapacity: number;
  }

  export module BuildingType {
    export var MOTHERSHIP: BuildingType = {
      id: 1,
      name: 'Mothership',
      cost: 0,
      validTerrain: TerrainType.BARE,
      constructionDays: 0,
      oreProduction: 1,
      miningCapacity: 5,
      techCapacity: 5,
      opsCapacity: 5
    };

    export var MINE: BuildingType = {
      id: 1,
      name: 'Mine',
      cost: 600,
      validTerrain: TerrainType.VEIN,
      constructionDays: 10,
      oreProduction: 10,
      miningCapacity: 5,
      techCapacity: 0,
      opsCapacity: 0
    };

     export var LAB: BuildingType = {
      id: 2,
      name: 'Lab',
      cost: 200,
      validTerrain: TerrainType.BARE,
      constructionDays: 5,
      oreProduction: 0,
      miningCapacity: 0,
      techCapacity: 5,
      opsCapacity: 0
    };

    export var OPS: BuildingType = {
      id: 3,
      name: 'Ops Center',
      cost: 200,
      validTerrain: TerrainType.BARE,
      constructionDays: 5,
      oreProduction: 0,
      miningCapacity: 0,
      techCapacity: 0,
      opsCapacity: 5
    };

    export var TUBE: BuildingType = {
      id: 4,
      name: 'Tube',
      cost: 50,
      validTerrain: TerrainType.BARE,
      constructionDays: 1,
      oreProduction: 0,
      miningCapacity: 0,
      techCapacity: 0,
      opsCapacity: 0
    };

    export var SPACEBANK: BuildingType = {
      id: 5,
      name: 'Spacebank',
      cost: 1000,
      validTerrain: TerrainType.BARE,
      constructionDays: 5,
      oreProduction: 0,
      miningCapacity: 0,
      techCapacity: 0,
      opsCapacity: 0
    };

    export var buildableBuildings = [ MINE, LAB, OPS, TUBE, SPACEBANK ];
  }
}

File: libgdx2/miner/core/src/org/quach/miner/engine/BuildingType.java
package org.quach.miner.engine;

import com.google.common.collect.ImmutableMap;

import java.util.Map;

public enum BuildingType {
  MOTHERSHIP(
    1,
    "Mothership",
    0,
    TerrainType.BARE,
    0,
    ImmutableMap.<Integer, Integer>of()
  ),

  MINE(
    2,
    "Mine",
    1000,
    TerrainType.VEIN,
    10,
    ImmutableMap.of(7, 1)
  ),

  REACTOR(
    3,
    "Reactor",
    1000,
    TerrainType.BARE,
    5,
    ImmutableMap.of(4, 1, 7, 1)
  ),

  CANTEEN(
    4,
    "Canteen",
    500,
    TerrainType.BARE,
    5,
    ImmutableMap.of(3, 1, 4, 1)
  ),

  TUBE(
    5,
    "Tube",
    100,
    TerrainType.BARE,
    1,
    ImmutableMap.<Integer, Integer>of()
  ),

  SPACEBANK(
    6,
    "Spacebank",
    1000,
    TerrainType.BARE,
    5,
    ImmutableMap.of(6, 100)
  ),

  QUARTERS(
    7,
    "Quarters",
    500,
    TerrainType.BARE,
    5,
    ImmutableMap.of(2, 1, 3, 1)
  );

  public final int id;
  public final String name;
  public final int cost;
  public final TerrainType buildableType;
  public final int constructionTime;
  public final Map<Integer, Integer> antiProximity;

  BuildingType(
    final int id,
    final String name,
    final int cost,
    final TerrainType buildableType,
    final int constructionTime,
    final Map<Integer, Integer> antiProximity
  ) {
    this.id = id;
    this.name = name;
    this.cost = cost;
    this.buildableType = buildableType;
    this.constructionTime = constructionTime;
    this.antiProximity = antiProximity;
  }

  public static final BuildingType[] BUILDABLE_BUILDINGS = { MINE, REACTOR, QUARTERS, CANTEEN, TUBE, SPACEBANK };
}
